File: Day 2/day_2.py
f = open("./input.txt", 'r')

file = f.readlines()

valid_passwords_part1 = 0
valid_passwords_part2 = 0

for i in range(len(file)):
    file[i] = file[i].replace("\n", '')
    temp = file[i].split('-')
    temp1 = temp[1].split(' ')
    p = [temp[0], temp1[0], temp1[1].replace(':', ''), temp1[2]]

    repetitions = 0

    # part 1: count the number of repetitions of the letter (p[2]) on the password (p[3]),
    # and check if it pass the requirements (p[0] <= n <= p[1])
    for l in p[3]:
        if(l == p[2]):
            repetitions += 1
    if repetitions >= int(p[0]) and repetitions <= int(p[1]):
        valid_passwords_part1 += 1

    # part 2: check if the letter is on (only) one of the indices -1 (XOR operation)
    if(bool(p[3][int(p[0]) - 1] == p[2]) ^ bool(p[3][int(p[1]) - 1] == p[2])):
        valid_passwords_part2 += 1

f.close()

print(f"part 1: {valid_passwords_part1} \npart 2: {valid_passwords_part2}")

File: Day 2/day_2.cpp
#include <iostream>
#include <fstream>
#include <string>
#include <vector>
#include <typeinfo>

using namespace std;

int main()
{
    fstream f;
    f.open("input.txt");
    string l;
    vector<string> passwords;

    int validPasswordsPt1 = 0;
    int validPasswordsPt2 = 0;

    //while there is a new line with a stringm push it (as a int) to 'report'
    while (getline(f, l))
    {
        passwords.push_back(l);
    }
    f.close();

    for (int i = 0; i < passwords.size(); i++)
    {
        string temp[4];
        bool i0 = true, i1 = true, i2 = true;
        int repetitions = 0;

        //Split the string (there must be an easier way to do this...)
        for (int j = 0; j < passwords[i].size(); j++)
        {
            if (passwords[i][j] == '-')
            {
                i0 = false;
                continue;
            }
            else if (i0)
            {
                temp[0] += passwords[i][j];
            }
            else if (i1 && passwords[i][j] == ' ')
            {
                i1 = false;
                continue;
            }
            else if (i1)
            {
                temp[1] += passwords[i][j];
            }
            else if (i2 && passwords[i][j] == ' ')
            {
                i2 = false;
                continue;
            }
            else if (i2)
            {
                if (passwords[i][j] != ':')
                    temp[2] += passwords[i][j];
            }
            else
            {
                temp[3] += passwords[i][j];
            }
        }

        // Part 1:
        for (int k = 0; k < temp[3].size(); k++)
        {
            if (temp[3][k] == temp[2][0])
            {
                repetitions++;
            }
        }
        if (repetitions >= stoi(temp[0]) && repetitions <= stoi(temp[1]))
        {
            validPasswordsPt1++;
        }

        // Part 2:
        if ((temp[3][stoi(temp[0]) - 1] == temp[2][0] && !(temp[3][stoi(temp[1]) - 1] == temp[2][0])) ||
            (temp[3][stoi(temp[1]) - 1] == temp[2][0] && !(temp[3][stoi(temp[0]) - 1] == temp[2][0])))
        {
            validPasswordsPt2++;
        }
    }

    cout << "Part 1: " << validPasswordsPt1 << "\nPart2: " << validPasswordsPt2 << endl;

    return 0;
}
